src/cdk/collections/selection-model.ts

```diff
@@ -76,7 +76,17 @@ export class SelectionModel<T> {
    */
   deselect(...values: T[]): boolean | void {
     this._verifyValueAssignment(values);
-    values.forEach(value => this._unmarkSelected(value));
+    if (this.compareWith) {
+      values.forEach(v1 => {
+        this._selection.forEach(v2 => {
+          if (this.compareWith?.call(undefined, v1, v2)) {
```

Comment:
Shouldn't this logic already be covered inside `_unmarkSelected`? It goes through each value and checks it with `isSelected` which will take the `compareWith` function into account.

Reply:
Yes, the isSelected does that, but the objective is to deselect by utilizing the actual reference, as the deselection process employs this._selection.delete(value) internally, which only works when the actual object reference is provided. (variable v2)

Reply:
I see. In that case I think that we should make the change in `_markSelected` and `_unmarkSelected`.

Reply:
I got it, I can work on that on this PR.

Reply:
It turns out that `_markSelected` does not need to be modified since it checks that a value isn't already selected before marking it selected. `if (!this.isSelected(value))`.

I created a private function `_getComparableValue` that could be reused (or even make public if there are some use-cases for it) in order to get a comparable selected value using the provided `compareWith` function.

Reply:
I end up creating the `_getConcreteValue` that returns  any comparable selected value, or returns the value passed as a parameter.
`_getConcreteValue` is called before selection / deselection.